### DISCUSSION TITLE
changing mass of body during run time?

### DISCUSSION THREAD

**Discussion by yoshymoo** [jrouwe/JoltPhysics#1323] [Q&A] [answered]
Hello!
I understand that we can specify the mass of a body during initialisation. But is it possible to change the mass of a body during run time? I'm looking at the ModifyMassTest sample and I know we can do so using the contact setting. But what if there's no contact?

**Comment by jrouwe** [comment]
Use `Body::GetMotionProperties` to get the `MotionProperties` object, then use `MotionProperties::SetMassProperties` or `MotionProperties::SetInverseMass` / `MotionProperties::SetInverseInertia`. Note that if you change mass, then normally you should change the inertia too (in the MotionProperties header there's more information about this).

**Reply by yoshymoo** [reply]
Hello! Thank you for replying. I am currently using the SetMassProperties function, here is how I'm using it:
```
JPH::BodyLockWrite lock(PHYSICS->physics_system.GetBodyLockInterface(), m_BodyID);
if (lock.Succeeded()) {
	JPH::MotionProperties* motionProperties = lock.GetBody().GetMotionProperties();
	JPH::MassProperties massProperties;
	//massProperties.mMass = 10.f;
	massProperties.ScaleToMass(10.f);
	motionProperties->SetMassProperties(JPH::EAllowedDOFs::All, massProperties);
}
```

I'm trying not to touch the inertia so I'm letting Jolt do the math. When I print out `lock.GetBody().GetMotionProperties()->GetInverseMass()`, it does give me the new mass which is 10.f, but it doesn't seem like it's doing anything...? I'm trying to adjust the mass for my object movement. 

Perhaps I am initialising wrongly? This is an example of how I am initialising my objects (I'm using scaled shape for all objects):
```
JPH::Ref<JPH::BoxShape> boxShape = new JPH::BoxShape(JPH::Vec3(1.f, 1.f, 1.f));
JPH::Ref<JPH::Shape> scaledShape = new JPH::ScaledShape(boxShape, JPH::Vec3(scale.x / 2.f, scale.y / 2.f, scale.z / 2.f));
JPH::BodyCreationSettings body(scaledShape, JPH::RVec3Arg(finalPos.x, finalPos.y, finalPos.z), placeHolder, eType, CollisionUtils::Layers::MOVING);
body.mOverrideMassProperties = JPH::EOverrideMassProperties::CalculateInertia;
body.mMassPropertiesOverride.mMass = mass;m_BodyID = PHYSICS->body_interface->CreateAndAddBody(body, activateBody);
```

**Comment by jrouwe** [comment]
This should work:

```
JPH::BodyLockWrite lock(PHYSICS->physics_system.GetBodyLockInterface(), m_BodyID);
if (lock.Succeeded()) {
	JPH::MotionProperties* motionProperties = lock.GetBody().GetMotionProperties();
	JPH::MassProperties massProperties = lock.GetBody().GetShape()->GetMassProperties();
	massProperties.ScaleToMass(10.f);
	motionProperties->SetMassProperties(JPH::EAllowedDOFs::All, massProperties);
}
```

But `GetMassProperties` can be expensive for complex shapes. I've added a new helper function `ScaleToMass` that makes it really simple to update the mass: #1326

**Comment by yoshymoo** [comment]
I'm a little confused as the mass doesn't seem to be doing anything for me. :(

Let's say I have Box A that I set to 100.f mass and Box B 1.f mass. When I use the add force function for both boxes, both boxes are accelerating at the same time. If I understand this correctly, the box with a larger mass will have a smaller acceleration, right? I'm also thinking that there should also be a difference in collision response if I collide both box A and box B together. Please take a look at my screen recording. 

https://github.com/user-attachments/assets/a9926b06-a7d6-40f9-99f2-fce094d70135

I'm using this function currently 
```
JPH::BodyLockWrite lock(PHYSICS->physics_system.GetBodyLockInterface(), m_BodyID);
if (lock.Succeeded()) {
	JPH::MotionProperties* motionProperties = lock.GetBody().GetMotionProperties();
	JPH::MassProperties massProperties = lock.GetBody().GetShape()->GetMassProperties();
	massProperties.ScaleToMass(10.f);
	motionProperties->SetMassProperties(JPH::EAllowedDOFs::All, massProperties);
}
```
and `float currentMass = 1.0f / lock.GetBody().GetMotionProperties()->GetInverseMass();` to retrieve the body's current mass.

**Reply by jrouwe** [reply, marked as answer]
> If I understand this correctly, the box with a larger mass will have a smaller acceleration, right?

Correct if you're using `BodyInterface/Body::AddForce` they will get different accelerations. If you're expecting the boxes to have different acceleration due to gravity then no, they will accelerate in the same way.

> I'm also thinking that there should also be a difference in collision response if I collide both box A and box B together.

Yes, this should also be the case. The heavy box should be far less affected.

If I change the `HeavyOnLightTest` sample to:

```
void HeavyOnLightTest::Initialize()
{
	// Floor
	CreateFloor();

	Ref<BoxShape> box = new BoxShape(Vec3::sReplicate(5));

	for (int i = 1; i <= 10; ++i)
	{
		Body *b1 = mBodyInterface->CreateBody(BodyCreationSettings(box, RVec3(-75.0f + i * 15.0f, 10.0f, 0.0f), Quat::sIdentity(), EMotionType::Dynamic, Layers::MOVING));
		mBodyInterface->AddBody(b1->GetID(), EActivation::Activate);
		MassProperties m = b1->GetShape()->GetMassProperties();
		m.ScaleToMass(1000.0f);
		b1->GetMotionProperties()->SetMassProperties(EAllowedDOFs::All, m);

		Body *b2 = mBodyInterface->CreateBody(BodyCreationSettings(box, RVec3(-75.0f + i * 15.0f, 30.0f, 0.0f), Quat::sIdentity(), EMotionType::Dynamic, Layers::MOVING));
		mBodyInterface->AddBody(b2->GetID(), EActivation::Activate);
		m = b1->GetShape()->GetMassProperties();
		m.ScaleToMass(5000.0f * i);
		b2->GetMotionProperties()->SetMassProperties(EAllowedDOFs::All, m);

		b2->AddForce(1.0e7f * Vec3::sAxisZ());
	}
}
```

Then both scenarios you describe just work. So I'm not sure what's different in your setup.

**Reply by yoshymoo** [reply]
Well, it turns out that I needed to set the mass again after changing the scale of the bodies :) 

Thank you so much!
